src/approaches/nsrt_rl_approach.py

```diff
@@ -66,16 +84,121 @@ def learn_from_interaction_results(
+
+        # For each training task, compute the experience data for each _Option
+        # we see used in that training task.
+        for i in range(len(self._train_tasks)):
+            plan = self._train_task_to_option_plan[i]
+            traj = self._train_task_to_online_traj[i]
+
+            curr_option_idx = 0
+            curr_option = plan[curr_option_idx]
+            parent_option = cast(_LearnedNeuralParameterizedOption,
+                                 curr_option.parent)
+            if parent_option not in option_to_data:
+                option_to_data[parent_option] = []
+            actions = ((i, a) for i, a in enumerate(traj.actions))
+            experience = []
+
+            # Loop through the trajectory and compute the experience tuples for
+            # each _Option in the plan.
+            for state, next_state in zip(traj.states[:-1], traj.states[1:]):
+                idx, action = next(actions)
+                state_features = state.vec(curr_option.objects)
+                relative_param = parent_option.get_rel_option_param_from_state(
+                    state,
+                    curr_option.memory,
+                    curr_option.objects,
+                )
+                # The RLOptionLearner will need the input to the option's
+                # regressor for training.
+                input_vec = cast(
+                    Array, np.hstack(([1.0], state_features, relative_param)))
+
+                # Add pos_reward if we got within epsilon of the option's
+                # subgoal, otherwise we add neg_reward. To do this, we can check
+                # the relative parameter for the next state.
+                next_rel_param = parent_option.get_rel_option_param_from_state(
+                    next_state,
+                    curr_option.memory,
+                    curr_option.objects,
+                )
+                reward: int = 0
+                if np.allclose(next_rel_param, 0, atol=self._reward_epsilon):
+                    reward += self._pos_reward
+                else:
+                    reward += self._neg_reward
+
+                experience.append(
+                    (state, input_vec, action, reward, next_state))
+
+                # Store experience data in two cases: (1) the next state is a
+                # terminal state for the current option and (2) the current
+                # option did not reach its terminal state but had a sufficient
+                # number of steps to reach it so that it is reasonable to assign
+                # it a reward.
+                terminate = parent_option.effect_based_terminal(
+                    next_state, curr_option.objects)
+                had_sufficient_steps = next_state.allclose(
+                    traj.states[-1]) and (
+                        CFG.max_num_steps_interaction_request -
+                        (idx + 1) > CFG.valid_reward_steps_threshold)
+                if terminate:
```

Comment:
as part of the suggested refactor into a helper function, it may also be good to have another smaller helper function that is responsible for processing the trajectory for a single option, until that option is done.